src/main/java/io/vertx/core/net/impl/URIDecoder.java

```diff
@@ -0,0 +1,126 @@
+/*
+ * Copyright (c) 2011-2017 Contributors to the Eclipse Foundation
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+ * which is available at https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+ */
+
+package io.vertx.core.net.impl;
+
+import io.netty.util.CharsetUtil;
+
+/**
+ * @author <a href="mailto:plopes@redhat.com">Paulo Lopes</a>
+ */
+public final class URIDecoder {
+
+  private URIDecoder() {
+    throw new RuntimeException("Static Class");
+  }
+
+  /**
+   * Decodes a segment of an URI encoded by a browser.
+   *
+   * The string is expected to be encoded as per RFC 3986, Section 2. This is the encoding used by JavaScript functions
+   * encodeURI and encodeURIComponent, but not escape. For example in this encoding, é (in Unicode U+00E9 or in
+   * UTF-8 0xC3 0xA9) is encoded as %C3%A9 or %c3%a9.
+   *
+   * Plus signs '+' will be handled as spaces and encoded using the default JDK URLEncoder class.
+   *
+   * @param s string to decode
+   *
+   * @return decoded string
+   */
+  public static String decodeURIComponent(String s) {
+    return decodeURIComponent(s, true);
+  }
+
+  /**
+   * Decodes a segment of an URI encoded by a browser.
+   *
+   * The string is expected to be encoded as per RFC 3986, Section 2. This is the encoding used by JavaScript functions
+   * encodeURI and encodeURIComponent, but not escape. For example in this encoding, é (in Unicode U+00E9 or in
+   * UTF-8 0xC3 0xA9) is encoded as %C3%A9 or %c3%a9.
+   *
+   * @param s string to decode
+   * @param plus weather or not to transform plus signs into spaces
+   *
+   * @return decoded string
+   */
+  public static String decodeURIComponent(String s, boolean plus) {
+    if (s == null) {
+      return null;
+    }
+
+    final int size = s.length();
+    boolean modified = false;
+    int i;
+    for (i = 0; i < size; i++) {
+      final char c = s.charAt(i);
+      if (c == '%' || (plus && c == '+')) {
+        modified = true;
+        break;
+      }
+    }
+    if (!modified) {
+      return s;
+    }
+    final byte[] buf = s.getBytes();
```

Comment:
Use an encoding when getting the bytes otherwise it will use the default platform charset

Reply:
Actually since we only write in buff, we should create a `byte[]`, using `getBytes()` will result in unnecessary CPU usage.

Reply:
I reuse the generated `byte[]` from the `i` where we detected the first `%` char. So I could either create a new `byte[]` and loop over the string to copy bytes (and i need bytes not chars) or let this do some extra work by copying the bytes and waste the decoded bytes from `i` to the end...